### ISSUE TITLE
[Feature Request]: HTMLElement to NativeImage

### ISSUE THREAD

**Issue by Awendel** [electron/electron#35394]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Currently there is no way in HTML to get the underlying pixelData of an HTMLElement (unless one uses canvas).

This is partially by design in order to protect users / stop fingerprinting etc.

Electron offers a partial solution currently via taking a Screenshot of a given webContent.
However there is 2 problems with that:
(1) Would have to manually crop if only subset of page desired
(2) Alpha / opacity data is lost (since it assumes every pixel to have alpha of 1.0

### Proposed Solution

Since the underlying Chromium exposes Rasterizations APIs and Buffers for pixel- / layerData, it would be trivial to open this up as a feature to be used in Electron.

This would have to of course happen from an elevated process (like preRender).

The following use cases could be supported:
- enable rasterization features, such as exporting content to jpg / png / pdf
- caching previews of content without having to re-render everything

### Alternatives Considered

The only alternative is using the "html2canvas" library.

But thie library is not pixelPerfect, has issues with especially more modern APIs (e.g. backdrop-filter, ...)
and is itself very large and slow.

### Additional Information

Since all the logic is already inbuilt into Chromium, it wouldn't take too much effort to bring it in Electron.

Would also be a great unique selling point for Electron, since we can open up features that are otherwise too "risky" for a normal browser to implement, yet are technically already available under the hood.

**Comment by Prinzhorn** [comment]
How would an API look like? This would be impossible with `contextIsolation` (you cannot pass an `HTMLElement` around). Your first problem with cropping is could be solved by passing `rect` to `capturePage`, e.g. via `getBoundingClientRect` which can be safely obtained in the renderer and serialized. What's left is that you seem to have issues with transparency not being conserved. I assume because you want to capture only a given element without things like the background color of the `body` being rendered as well? Or in the case of `getBoundingClientRect` you wouldn't want elements that overlap it (e.g. position: fixed) to also be capture?

**Comment by Awendel** [comment]
Yes, the main issue would be that transparency would never be preversed, which would be a crucial feature for exporting png with transparency. Currently the only way to do this to support exporting HTML Content with transparency would be trying to create a custom render using Canvas or WebGL - where you'd then lose a lot of functionality around multiLine text and would have to recreate properties like backdropfilter manually.

API would depend on the context and where I could be integrated (either from Node js runtime or pre-render).

But if passing on references directly to an HTMLElement wouldn't be possible, one could always have the option to do it via a querySelector / CSS selector like syntax (so the input param would be string instead of HTMLElement).

**Comment by Awendel** [comment]
The alternative of course would be to extend the existing Screenshot API to offer an option to make the background / body transparent.
If this is not natively possible, there is a way to "hack" this behaviour:

Take 3 seperate Screenshots, at each stage setting the background Color of body / webContent to:
• pure red
• pure green
• pure blue

then by comparing each pixel value how it differs with each background, one can infer the underlying color as well as the alpha. I'd be happy to supply the Math formulas / code for that algorithm if required

**Comment by Prinzhorn** [comment]
> If this is not natively possible, there is a way to "hack" this behaviour:

I think for most if not all websites a simple hack that works rn is something like `* {visibility: none !important}` and set the element you want to screenshot to `.screenshot, .screenshot * {visibility: visible !important}`. This can be done before taking the screenshot and removed after taking it. So it would likely cause a brief flicker. The logic regarding `visibility` and how to apply and remove it can probably optimized, this is just an idea.

**Comment by Awendel** [comment]
That is a good idea for ensuring no overlaps etc, but doesn't solve the missing alpha data issue.

I guess there is 2 more features where I'm not certain the existing Screenshot API suffices:

(1) Would it be possible to take a screenshot of an invisible BrowserView / window - so I could take the screenshot "off-thread" and user wouldn't see the flickering?

(2) Say I want to export a very large board with a resolution of something like 6000x4000px - in a normal browserWindow it would be clipped of course. But would I be able to temporarily set the (invisible) BrowserView to that resolution without any problems?

**Comment by Prinzhorn** [comment]
> but doesn't solve the missing alpha data issue.

Can you provide a REPL? From what I understand PNGs captured with `capturePage` do have transparency if the there is no background.

**Comment by Awendel** [comment]
My apologies, I based my assumptions only on the documentation and assumed it would take a screenshot similar to normal screen capture. 

Perhaps a small section in the docs regarding transparency behaviour would be useful (e.g. would the browserView / window have to be transparent / and / or the body of the html)?

If taking screenshots of "invisible" browserViews (e.g. not appended to BrowserView) is supported, then this issue can definitely be closed, and the behaviour might be best implemented / emulated in an external 3rd party library.

**Comment by Prinzhorn** [comment]
It would be great if you can create a Fiddle that answers all these questions. This would also help others that come across this issue in the future. https://github.com/electron/fiddle

> Perhaps a small section in the docs regarding transparency behaviour would be useful (e.g. would the browserView / window have to be transparent / and / or the body of the html)?

You can edit the docs https://www.electronjs.org/docs/latest/api/web-contents#contentscapturepagerect here https://github.com/electron/electron/edit/main/docs/api/web-contents.md . There's an "Edit this page" link at the end of every docs page.

**Comment by Awendel** [comment]
Alright, I'll write a little test case over the few days when I find time and report back to see if any behaviour is missing and will annotate docs based on findings.

Until then we can leave the issue open and check again when we know more.

**Comment by Awendel** [comment]
Wrote a little Fiddle to answer some of the questions:
(1) "offscreen rendering" is indeed possible by using a BrowserView with show:false
(2) Can be set to any arbitrary resolution (even 9000x9000) etc
(3) transparency is possible with png screenshots if the backgroundColor of the browserWindow is transparent and transparent:true and the body in html is also transparent. When using jpg transparent regions become black.

This does indeed answer all questions.
I'll add a small guide to transparent screenshots in the documentation later or tomorrow.

The issue can be closed.

**Comment by Prinzhorn** [comment]
> The issue can be closed.

As the author you possess this power

**Comment by 1mtiyxz** [comment]
please assign this issue to me

**Comment by fabiospampinato** [comment]
This would be pretty useful, for many use cases. For example I'm drawing an xScope-like crosshair to measure things, the thing is my crosshair needs a screenshot of the app to work, but if the screenshot contains the crosshair then it doesn't work anymore, but if I had the crosshair just to take a clean screenshot the experience is not as nice. Really the solution would be taking a screenshot of a particular Node, so that I can exclude nodes outside of it.

**Comment by samuelmaddock** [comment]
This should be possible using the upcoming Element Capture API
https://developer.chrome.com/docs/web-platform/element-capture

**Comment by LaNsHoR** [comment]
This would be very helpful, my current use case I can't implent:

- I have a png image with some transparent pixels.
- I want to trigger pointer events (click, enter, leave) only for the non-transparent pixels of the image.
- The image has some CSS transform / animations (scale, translate, rotate) so get the image data and put it on a canvas does not work (we lost the transform / animation so the pixel coordinates are unkown).
- The image has other content behind so full viewport screentshoot won't help as I can't discriminate between the image pixels and other elements pixels.

Ideally I would love to have a CSS mask for events, but as the spec has nothing like that offering an API in electron would help a lot.

**Comment by LaNsHoR** [comment]
The recommendation of https://github.com/electron/electron/issues/35394#issuecomment-2134081422 doesn't seem a valid solution in multiple use cases, like the one I described in https://github.com/electron/electron/issues/35394#issuecomment-2358029069

Some reasons are:
- Element capture does NOT capture alpha channel.
- Element capture does NOT capture screen cordinates of the element (like CSS transforms or element animations).
